Mathlib/NumberTheory/ModularForms/DedekindEta.lean

```diff
@@ -0,0 +1,143 @@
+-/
+
+open UpperHalfPlane TopologicalSpace Set MeasureTheory intervalIntegral
+ Metric Filter Function Complex
+
+open scoped Interval Real NNReal ENNReal Topology BigOperators Nat
+
+local notation "𝕢" => Periodic.qParam
+
+local notation "ℍₒ" => complexUpperHalfPlane
+
+/-- The q term inside the product defining the eta function. It is defined as
+`eta_q n z = e ^ (2 π i (n + 1) z)`. -/
+noncomputable abbrev eta_q (n : ℕ) (z : ℂ) := (𝕢 1 z) ^ (n + 1)
+
+lemma eta_q_eq_cexp (n : ℕ) (z : ℂ) : eta_q n z = cexp (2 * π * Complex.I * (n + 1) * z) := by
+  simp [eta_q, Periodic.qParam, ← Complex.exp_nsmul]
+  ring_nf
+
+lemma eta_q_eq_pow (n : ℕ) (z : ℂ) : eta_q n z = cexp (2 * π * Complex.I * z) ^ (n + 1) := by
+  simp [eta_q, Periodic.qParam]
+
+lemma one_add_eta_q_ne_zero (n : ℕ) (z : ℍ) : 1 - eta_q n z ≠ 0 := by
+  rw [eta_q_eq_cexp, sub_ne_zero]
+  intro h
+  have := norm_exp_two_pi_I_lt_one ⟨(n + 1) • z, by
+    have : 0 < (n + 1 : ℝ) := by linarith
+    simpa [this] using z.2⟩
+  simp [← mul_assoc, ← h] at *
+
+/-- The product term in the eta function, defined as `∏' 1 - q ^ (n + 1)` for `q = e ^ 2 π i z`. -/
+noncomputable abbrev etaProdTerm (z : ℂ) := ∏' (n : ℕ), (1 - eta_q n z)
+
+local notation "ηₚ" => etaProdTerm
+
+/-- The eta function, whose value at z is `q^ 1 / 24 * ∏' 1 - q ^ (n + 1)` for `q = e ^ 2 π i z`. -/
+noncomputable def ModularForm.eta (z : ℂ) := 𝕢 24 z * ηₚ z
+
+local notation "η" => ModularForm.eta
+
+open ModularForm
+
+theorem Summable_eta_q (z : ℍ) : Summable fun n ↦ ‖-eta_q n z‖ := by
+  simp [eta_q, eta_q_eq_pow, summable_nat_add_iff 1, norm_exp_two_pi_I_lt_one z]
+
+lemma hasProdLocallyUniformlyOn_eta : HasProdLocallyUniformlyOn (fun n a ↦ 1 - eta_q n a) ηₚ ℍₒ:= by
+  simp_rw [sub_eq_add_neg]
+  apply hasProdLocallyUniformlyOn_of_forall_compact complexUpperHalPlane_isOpen
+  intro K hK hcK
+  by_cases hN : K.Nonempty
+  · have hc : ContinuousOn (fun x ↦ ‖cexp (2 * π * Complex.I * x)‖) K := by fun_prop
+    obtain ⟨z, hz, hB, HB⟩ := hcK.exists_sSup_image_eq_and_ge hN hc
+    apply (Summable_eta_q ⟨z, by simpa using (hK hz)⟩).hasProdUniformlyOn_nat_one_add hcK
+    · filter_upwards with n x hx
+      simpa only [eta_q, eta_q_eq_pow n x, norm_neg, norm_pow, coe_mk_subtype,
+          eta_q_eq_pow n (⟨z, hK hz⟩ : ℍ)] using
+          pow_le_pow_left₀ (by simp [norm_nonneg]) (HB x hx) (n + 1)
+    · simp_rw [eta_q, Periodic.qParam]
+      fun_prop
+  · rw [hasProdUniformlyOn_iff_tendstoUniformlyOn]
+    simpa [not_nonempty_iff_eq_empty.mp hN] using tendstoUniformlyOn_empty
+
+theorem etaProdTerm_ne_zero (z : ℍ) : ηₚ z ≠ 0 := by
+  simp only [etaProdTerm, eta_q, ne_eq]
+  refine tprod_one_add_ne_zero_of_summable z (f := fun n x ↦ -eta_q n x) ?_ ?_
+  · refine fun i x ↦ by simpa using one_add_eta_q_ne_zero i x
+  · intro x
+    simpa [eta_q, ← summable_norm_iff] using Summable_eta_q x
+
+/-- Eta is non-vanishing on the upper half plane. -/
+lemma eta_ne_zero_on_UpperHalfPlane (z : ℍ) : η z ≠ 0 := by
```

Comment:
Here the subset, not the subtype, is appropriate since `η` was defined as a function on `ℂ` (& same in preceding lemma and various other places in this PR)